spinalcordtoolbox/aggregate_slicewise.py

```diff
@@ -44,8 +44,173 @@ def __init__(self, id, name, filename=None, map_cluster=None):
+                                  (y - np.dot(x, beta_0))))
+    return beta[0], beta
+
+
+def func_ml(data, mask, map_clusters=None):
+    """
+    Compute maximum likelihood (ML) for the first label of mask.
+    :param data: nd-array: input data
+    :param mask: (n+1)d-array: input mask. Note: this mask should include ALL labels to satisfy the necessary condition
+    for ML-based estimation, i.e., at each voxel, the sum of all labels (across the last dimension) equals the
+    probability to be inside the tissue. For example, for a pixel within the spinal cord, the sum of all labels should
+    be 1.
+    :return: float: beta corresponding to the first label
+    """
+    # TODO: support weighted least square
+    # reshape as 1d vector (for data) and 2d vector (for mask)
+    n_vox = functools.reduce(operator.mul, data.shape, 1)
+    # ML estimation:
+    #   y: measurements vector (to which weights are applied)
+    #   x: linear relation between the measurements y
+    #   beta [nb_labels] = (Xt . X)^(-1) . Xt . y: The estimated metric value in each label
+    y = np.reshape(data, n_vox)  # [nb_vox x 1]
+    x = np.reshape(mask, (n_vox, mask.shape[mask.ndim-1]))
+    beta = np.dot(np.linalg.pinv(np.dot(x.T, x)), np.dot(x.T, y))
+    return beta[0], beta
+
+
+def func_std(data, mask=None, map_clusters=None):
+    """
+    Compute standard deviation
+    :param data: nd-array: input data
+    :param mask: (n+1)d-array: input mask
+    :param map_clusters: not used
+    :return:
+    """
+    # Check if mask has an additional dimension (in case it is a label). If so, select the first label
+    if mask.ndim == data.ndim + 1:
+        mask = mask[..., 0]
+    average, _ = func_wa(data, np.expand_dims(mask, axis=mask.ndim))
+    variance = np.average((data - average) ** 2, weights=mask)
+    return math.sqrt(variance), None
+
+
+def func_sum(data, mask=None, map_clusters=None):
+    """
+    Compute sum
+    :param data: nd-array: input data
+    :param mask: not used
+    :param map_clusters: not used
+    :return:
+    """
+    # Check if mask has an additional dimension (in case it is a label). If so, select the first label
+    return np.sum(data), None
+
+
+def func_wa(data, mask=None, map_clusters=None):
+    """
+    Compute weighted average
+    :param data: nd-array: input data
+    :param mask: (n+1)d-array: input mask
+    :param map_clusters: not used
+    :return:
+    """
+    # Check if mask has an additional dimension (in case it is a label). If so, select the first label
+    if mask.ndim == data.ndim + 1:
+        mask = mask[..., 0]
+    return np.average(data, weights=mask), None
+
+
 def aggregate_per_slice_or_level(metric, mask=None, slices=[], levels=[], perslice=None, perlevel=False,
-                                 vert_level=None, group_funcs=(('MEAN', np.mean),), map_clusters=None):
+                                 vert_level=None, group_funcs=(('MEAN', func_wa),), map_clusters=None):
```

Comment:
Just curious: Why did you change this (`np.mean` --> `func_wa`)?

Reply:
good question. because the default `np.mean` was failing, due to recent changes in the function which expects two outputs, as can be seen [here](https://github.com/neuropoly/spinalcordtoolbox/blob/0469ed131fb580f01185d54fbda3ee86e9d37025/spinalcordtoolbox/aggregate_slicewise.py#L279). It never crashed before because we always give input to this function, but the default value would crash.